azure-quantum/azure/quantum/qiskit/provider.py

```diff
@@ -2,59 +2,91 @@
 # Copyright (c) Microsoft Corporation.
 # Licensed under the MIT License.
 ##
+
 try:
     from qiskit.providers import ProviderV1 as Provider
     from qiskit.providers.exceptions import QiskitBackendNotFoundError
+    from qiskit.providers import BackendV1 as Backend
+    from qiskit.exceptions import QiskitError
 except ImportError:
     raise ImportError(
-    "Missing optional 'qiskit' dependencies. \
+        "Missing optional 'qiskit' dependencies. \
 To install run: pip install azure-quantum[qiskit]"
     )
 
-from typing import Dict, Iterable
-from azure.quantum import Workspace
 
+from typing import Dict, List, Tuple, Type
+from azure.quantum import Workspace
+from azure.quantum.qiskit.backends.backend import AzureBackendBase
 from azure.quantum.qiskit.job import AzureQuantumJob
 from azure.quantum.qiskit.backends import *
+from itertools import groupby
+import warnings
+import inspect
+
+# Target ID keyword for parameter-free solvers
+PARAMETER_FREE = "parameterfree"
 
 QISKIT_USER_AGENT = "azure-quantum-qiskit"
 
 
 class AzureQuantumProvider(Provider):
     def __init__(self, workspace=None, **kwargs):
-        self._backends = None
         if workspace is None:
             workspace = Workspace(**kwargs)
 
         workspace.append_user_agent(QISKIT_USER_AGENT)
 
         self._workspace = workspace
+        self._backends = None
 
     def get_workspace(self) -> Workspace:
         return self._workspace
 
-    def _get_all_subclasses(self, cls):
-        all_subclasses = []
-        for subclass in cls.__subclasses__():
-            all_subclasses.append(subclass)
-            all_subclasses.extend(self._get_all_subclasses(subclass))
-        return all_subclasses
-
     def get_backend(self, name=None, **kwargs):
+        """Return a single backend matching the specified filtering.
+        Args:
+            name (str): name of the backend.
+            **kwargs: dict used for filtering.
+        Returns:
+            Backend: a backend matching the filtering.
+        Raises:
+            QiskitBackendNotFoundError: if no backend could be found or
+                more than one backend matches the filtering criteria.
         """
-        Return a single backend matching the specified filtering.
-        """
-        backends = self.backends(name, **kwargs)
+
+        backends = self.backends(name=name, **kwargs)
+
+        if len(backends) > 1:
+            def all_same(iterable):
+                group_iter = groupby(iterable)
+                return next(group_iter, True) and not next(group_iter, False)
+
+            # If all backends have the same name, filter for default backend
+            if all_same(backend.name() for backend in backends):
```

Comment:
Isn't it known at this point that all the backends in 'backends' have the same name? Given that we specify a name filter in line 58?

Reply:
The `name` defaults to `None`: `def get_backend(self, name=None, **kwargs):`  - so we don't filter on name unless it is supplied. Line 58, `backends = self.backends(name=name, **kwargs)`, forwards the `name` or `None` if unset to the `self.backends`.